File: mobile/lib/features/games/domain/okey101/okey101_engine.dart
import 'dart:math';

import 'okey101_models.dart';

enum Okey101MoveType { start, draw, takeDiscard, discard, open }

class Okey101Move {
  const Okey101Move._(this.type, {this.tileId});

  const Okey101Move.start() : this._(Okey101MoveType.start);
  const Okey101Move.draw() : this._(Okey101MoveType.draw);
  const Okey101Move.takeDiscard() : this._(Okey101MoveType.takeDiscard);
  const Okey101Move.open() : this._(Okey101MoveType.open);
  const Okey101Move.discard(String tileId)
      : this._(Okey101MoveType.discard, tileId: tileId);

  factory Okey101Move.fromJson(Map<String, dynamic> json) {
    final type = json['type']?.toString() ?? 'draw';
    switch (type) {
      case 'start':
        return const Okey101Move.start();
      case 'draw':
        return const Okey101Move.draw();
      case 'takeDiscard':
        return const Okey101Move.takeDiscard();
      case 'open':
        return const Okey101Move.open();
      case 'discard':
        return Okey101Move.discard(json['tileId']?.toString() ?? '');
      default:
        return const Okey101Move.draw();
    }
  }

  final Okey101MoveType type;
  final String? tileId;
}

class Okey101PlayerState {
  Okey101PlayerState({
    required this.userId,
    required this.displayName,
    required this.seat,
    List<Okey101Tile>? hand,
    this.opened = false,
    this.penalty = 0,
    this.connected = true,
    this.drewThisTurn = false,
  }) : hand = hand ?? [];

  final String userId;
  final String displayName;
  final int seat;
  final List<Okey101Tile> hand;
  bool opened;
  int penalty;
  bool connected;
  bool drewThisTurn;

  Okey101PlayerState copyWith({
    List<Okey101Tile>? hand,
    bool? opened,
    int? penalty,
    bool? drewThisTurn,
  }) {
    return Okey101PlayerState(
      userId: userId,
      displayName: displayName,
      seat: seat,
      hand: hand ?? this.hand,
      opened: opened ?? this.opened,
      penalty: penalty ?? this.penalty,
      connected: connected,
      drewThisTurn: drewThisTurn ?? this.drewThisTurn,
    );
  }
}

class Okey101GameState {
  Okey101GameState({
    required this.hostUserId,
    this.phase = 'waiting',
    this.status = 'waiting',
    this.openScoreTarget = 101,
    this.maxPlayers = 4,
    this.minPlayers = 2,
    this.indicator,
    this.okey,
    List<Okey101Tile>? deck,
    List<Okey101Tile>? discard,
    List<Okey101PlayerState>? players,
    this.turnIndex = 0,
    this.winnerId,
    this.winnerName,
    this.lastMove,
  })  : deck = deck ?? [],
        discard = discard ?? [],
        players = players ?? [];

  String phase;
  String status;
  final int openScoreTarget;
  final int maxPlayers;
  final int minPlayers;
  Okey101Tile? indicator;
  Okey101Tile? okey;
  List<Okey101Tile> deck;
  List<Okey101Tile> discard;
  List<Okey101PlayerState> players;
  int turnIndex;
  String? winnerId;
  String? winnerName;
  String? lastMove;
  final String hostUserId;
}

final _rng = Random();

List<Okey101Tile> _createDeck() {
  final tiles = <Okey101Tile>[];
  for (final color in [
    OkeyColor.red,
    OkeyColor.yellow,
    OkeyColor.blue,
    OkeyColor.black,
  ]) {
    for (var value = 1; value <= 13; value++) {
      tiles.add(Okey101Tile(id: '${color.name}-$value-a', color: color, value: value));
      tiles.add(Okey101Tile(id: '${color.name}-$value-b', color: color, value: value));
    }
  }
  tiles.add(
    const Okey101Tile(id: 'joker-fake-a', color: OkeyColor.joker, value: 0, isFakeJoker: true),
  );
  tiles.add(
    const Okey101Tile(id: 'joker-fake-b', color: OkeyColor.joker, value: 0, isFakeJoker: true),
  );
  tiles.shuffle(_rng);
  return tiles;
}

Okey101Tile _resolveOkey(Okey101Tile indicator) {
  if (indicator.isFakeJoker || indicator.color == OkeyColor.joker) {
    return const Okey101Tile(id: 'okey-wild', color: OkeyColor.red, value: 1);
  }
  final next = indicator.value == 13 ? 1 : indicator.value + 1;
  return Okey101Tile(
    id: 'okey-${indicator.color.name}-$next',
    color: indicator.color,
    value: next,
  );
}

int _handPoints(List<Okey101Tile> hand, Okey101Tile? okey) {
  return hand.fold(0, (sum, tile) => sum + tile.points(okey));
}

bool _hasBasicGroup(List<Okey101Tile> hand, Okey101Tile? okey) {
  final byColor = <String, List<int>>{};
  final byValue = <int, int>{};
  for (final tile in hand) {
    final wild = okey != null && tile.isWild(okey);
    final color = wild ? 'wild' : tile.color.name;
    final value = wild ? -1 : tile.value;
    if (color != 'wild') {
      byColor.putIfAbsent(color, () => []).add(value);
    }
    if (value > 0) {
      byValue[value] = (byValue[value] ?? 0) + 1;
    }
  }
  for (final values in byColor.values) {
    values.sort();
    var run = 1;
    for (var i = 1; i < values.length; i++) {
      if (values[i] == values[i - 1] + 1) {
        run++;
        if (run >= 3) return true;
      } else if (values[i] != values[i - 1]) {
        run = 1;
      }
    }
  }
  for (final count in byValue.values) {
    if (count >= 3) return true;
  }
  if (okey != null && hand.where((t) => t.isWild(okey)).length >= 2) {
    return true;
  }
  return false;
}

Okey101GameState createOkey101State(String hostUserId, String hostName) {
  return Okey101GameState(
    hostUserId: hostUserId,
    players: [
      Okey101PlayerState(
        userId: hostUserId,
        displayName: hostName,
        seat: 0,
      ),
    ],
  );
}

Okey101GameState addPlayer(
  Okey101GameState state,
  String userId,
  String displayName,
) {
  if (state.phase != 'waiting') {
    throw StateError('Oyun başlamış');
  }
  if (state.players.any((p) => p.userId == userId)) return state;
  if (state.players.length >= state.maxPlayers) {
    throw StateError('Oda dolu');
  }
  return Okey101GameState(
    hostUserId: state.hostUserId,
    phase: state.phase,
    status: state.status,
    openScoreTarget: state.openScoreTarget,
    maxPlayers: state.maxPlayers,
    minPlayers: state.minPlayers,
    indicator: state.indicator,
    okey: state.okey,
    deck: state.deck,
    discard: state.discard,
    players: [
      ...state.players,
      Okey101PlayerState(
        userId: userId,
        displayName: displayName,
        seat: state.players.length,
      ),
    ],
    turnIndex: state.turnIndex,
    winnerId: state.winnerId,
    winnerName: state.winnerName,
    lastMove: state.lastMove,
  );
}

Okey101GameState startGame(Okey101GameState state, String userId) {
  if (state.phase != 'waiting') throw StateError('Oyun zaten başladı');
  if (state.players.length < state.minPlayers) {
    throw StateError('En az ${state.minPlayers} oyuncu gerekli');
  }
  if (userId != state.hostUserId) {
    throw StateError('Yalnızca oda sahibi başlatabilir');
  }

  final deck = _createDeck();
  final indicator = deck.removeLast();
  final okey = _resolveOkey(indicator);

  final players = <Okey101PlayerState>[];
  for (var i = 0; i < state.players.length; i++) {
    final count = i == 0 ? 15 : 14;
    final hand = <Okey101Tile>[];
    for (var j = 0; j < count; j++) {
      hand.add(deck.removeLast());
    }
    players.add(state.players[i].copyWith(hand: hand, opened: false));
  }

  final discard = deck.isNotEmpty ? [deck.removeLast()] : <Okey101Tile>[];

  return Okey101GameState(
    hostUserId: state.hostUserId,
    phase: 'playing',
    status: 'playing',
    indicator: indicator,
    okey: okey,
    deck: deck,
    discard: discard,
    players: players,
    lastMove: 'Oyun başladı',
  );
}

Okey101GameState _applyPenalties(Okey101GameState state, String winnerId) {
  final players = state.players.map((p) {
    if (p.userId == winnerId) return p;
    return p.copyWith(penalty: p.penalty + _handPoints(p.hand, state.okey));
  }).toList();
  return Okey101GameState(
    hostUserId: state.hostUserId,
    phase: state.phase,
    status: state.status,
    openScoreTarget: state.openScoreTarget,
    maxPlayers: state.maxPlayers,
    minPlayers: state.minPlayers,
    indicator: state.indicator,
    okey: state.okey,
    deck: state.deck,
    discard: state.discard,
    players: players,
    turnIndex: state.turnIndex,
    winnerId: state.winnerId,
    winnerName: state.winnerName,
    lastMove: state.lastMove,
  );
}

Okey101GameState applyMove(
  Okey101GameState state,
  String userId,
  Okey101Move move,
) {
  if (move.type == Okey101MoveType.start) {
    return startGame(state, userId);
  }
  if (state.phase != 'playing') throw StateError('Oyun aktif değil');

  final playerIndex = state.players.indexWhere((p) => p.userId == userId);
  if (playerIndex < 0) throw StateError('Oyuncu değilsiniz');
  if (playerIndex != state.turnIndex) throw StateError('Sıra sizde değil');

  final player = state.players[playerIndex];
  final okey = state.okey;

  switch (move.type) {
    case Okey101MoveType.draw:
      if (player.drewThisTurn) throw StateError('Bu turda zaten çektiniz');
      if (state.deck.isEmpty) throw StateError('Deste bitti');
      final drawn = state.deck.removeLast();
      final players = [...state.players];
      players[playerIndex] = player.copyWith(
        hand: [...player.hand, drawn],
        drewThisTurn: true,
      );
      return Okey101GameState(
        hostUserId: state.hostUserId,
        phase: state.phase,
        status: state.status,
        openScoreTarget: state.openScoreTarget,
        maxPlayers: state.maxPlayers,
        minPlayers: state.minPlayers,
        indicator: state.indicator,
        okey: state.okey,
        deck: state.deck,
        discard: state.discard,
        players: players,
        turnIndex: state.turnIndex,
        lastMove: '${player.displayName} desteden çekti',
      );

    case Okey101MoveType.takeDiscard:
      if (player.drewThisTurn) throw StateError('Bu turda zaten çektiniz');
      if (state.discard.isEmpty) throw StateError('Atık yok');
      final taken = state.discard.removeLast();
      final takePlayers = [...state.players];
      takePlayers[playerIndex] = player.copyWith(
        hand: [...player.hand, taken],
        drewThisTurn: true,
      );
      return Okey101GameState(
        hostUserId: state.hostUserId,
        phase: state.phase,
        status: state.status,
        openScoreTarget: state.openScoreTarget,
        maxPlayers: state.maxPlayers,
        minPlayers: state.minPlayers,
        indicator: state.indicator,
        okey: state.okey,
        deck: state.deck,
        discard: state.discard,
        players: takePlayers,
        turnIndex: state.turnIndex,
        lastMove: '${player.displayName} atılan taşı aldı',
      );

    case Okey101MoveType.discard:
      if (!player.drewThisTurn && player.hand.length > 14) {
        throw StateError('Önce taş çekmelisiniz');
      }
      final tileId = move.tileId ?? '';
      final idx = player.hand.indexWhere((t) => t.id == tileId);
      if (idx < 0) throw StateError('Taş elinizde yok');
      final tile = player.hand[idx];
      final newHand = [...player.hand]..removeAt(idx);
      final discardPlayers = [...state.players];
      discardPlayers[playerIndex] = player.copyWith(hand: newHand);
      var next = Okey101GameState(
        hostUserId: state.hostUserId,
        phase: state.phase,
        status: state.status,
        openScoreTarget: state.openScoreTarget,
        maxPlayers: state.maxPlayers,
        minPlayers: state.minPlayers,
        indicator: state.indicator,
        okey: state.okey,
        deck: state.deck,
        discard: [...state.discard, tile],
        players: discardPlayers,
        turnIndex: state.turnIndex,
        lastMove: '${player.displayName} taş attı',
      );
      if (player.opened && newHand.isEmpty) {
        next = Okey101GameState(
          hostUserId: next.hostUserId,
          phase: 'finished',
          status: 'finished',
          openScoreTarget: next.openScoreTarget,
          maxPlayers: next.maxPlayers,
          minPlayers: next.minPlayers,
          indicator: next.indicator,
          okey: next.okey,
          deck: next.deck,
          discard: next.discard,
          players: next.players,
          turnIndex: playerIndex,
          winnerId: userId,
          winnerName: player.displayName,
          lastMove: '${player.displayName} okey attı!',
        );
        return _applyPenalties(next, userId);
      }
      final nextIndex = (state.turnIndex + 1) % state.players.length;
      final resetPlayers = next.players
          .map((p) => p.copyWith(drewThisTurn: false))
          .toList();
      return Okey101GameState(
        hostUserId: next.hostUserId,
        phase: next.phase,
        status: next.status,
        openScoreTarget: next.openScoreTarget,
        maxPlayers: next.maxPlayers,
        minPlayers: next.minPlayers,
        indicator: next.indicator,
        okey: next.okey,
        deck: next.deck,
        discard: next.discard,
        players: resetPlayers,
        turnIndex: nextIndex,
        winnerId: next.winnerId,
        winnerName: next.winnerName,
        lastMove: next.lastMove,
      );

    case Okey101MoveType.open:
      final points = _handPoints(player.hand, okey);
      if (points < state.openScoreTarget) {
        throw StateError('Açmak için en az ${state.openScoreTarget} puan gerekli');
      }
      if (!_hasBasicGroup(player.hand, okey)) {
        throw StateError('Geçerli per/çift yok');
      }
      final openPlayers = [...state.players];
      openPlayers[playerIndex] = player.copyWith(opened: true);
      return Okey101GameState(
        hostUserId: state.hostUserId,
        phase: state.phase,
        status: state.status,
        openScoreTarget: state.openScoreTarget,
        maxPlayers: state.maxPlayers,
        minPlayers: state.minPlayers,
        indicator: state.indicator,
        okey: state.okey,
        deck: state.deck,
        discard: state.discard,
        players: openPlayers,
        turnIndex: state.turnIndex,
        lastMove: '${player.displayName} 101 ile açtı ($points puan)',
      );

    case Okey101MoveType.start:
      return startGame(state, userId);
  }
}

Map<String, dynamic> publicView(Okey101GameState state, String viewerUserId) {
  return {
    'gameType': 'okey101',
    'phase': state.phase,
    'status': state.status,
    'openScoreTarget': state.openScoreTarget,
    'maxPlayers': state.maxPlayers,
    'minPlayers': state.minPlayers,
    'indicator': state.indicator == null
        ? null
        : {
            'id': state.indicator!.id,
            'color': state.indicator!.color.name,
            'value': state.indicator!.value,
            'isFakeJoker': state.indicator!.isFakeJoker,
          },
    'okey': state.okey == null
        ? null
        : {
            'id': state.okey!.id,
            'color': state.okey!.color.name,
            'value': state.okey!.value,
          },
    'deckCount': state.deck.length,
    'discardTop': state.discard.isEmpty
        ? null
        : {
            'id': state.discard.last.id,
            'color': state.discard.last.color.name,
            'value': state.discard.last.value,
            'isFakeJoker': state.discard.last.isFakeJoker,
          },
    'discardCount': state.discard.length,
    'turnUserId': state.players.isEmpty ? null : state.players[state.turnIndex].userId,
    'turnName': state.players.isEmpty ? null : state.players[state.turnIndex].displayName,
    'winnerId': state.winnerId,
    'winnerName': state.winnerName,
    'lastMove': state.lastMove,
    'hostUserId': state.hostUserId,
    'players': state.players.map((p) {
      final isMe = p.userId == viewerUserId;
      return {
        'userId': p.userId,
        'displayName': p.displayName,
        'seat': p.seat,
        'handCount': p.hand.length,
        'opened': p.opened,
        'penalty': p.penalty,
        'connected': p.connected,
        'isMe': isMe,
        'hand': isMe
            ? p.hand
                .map(
                  (t) => {
                    'id': t.id,
                    'color': t.color.name,
                    'value': t.value,
                    'isFakeJoker': t.isFakeJoker,
                  },
                )
                .toList()
            : [],
      };
    }).toList(),
    'myHandPoints': _handPoints(
      state.players
              .where((p) => p.userId == viewerUserId)
              .map((p) => p.hand)
              .firstOrNull ??
          const [],
      state.okey,
    ),
  };
}

File: api/src/lib/okey101Engine.ts
export type OkeyColor = "red" | "yellow" | "blue" | "black";

export interface OkeyTile {
  id: string;
  color: OkeyColor | "joker";
  value: number;
  isFakeJoker?: boolean;
}

export interface Okey101Player {
  userId: string;
  displayName: string;
  seat: number;
  hand: OkeyTile[];
  opened: boolean;
  penalty: number;
  connected: boolean;
  drewThisTurn?: boolean;
}

export interface Okey101State {
  gameType: "okey101";
  phase: "waiting" | "playing" | "finished";
  status: string;
  openScoreTarget: number;
  maxPlayers: number;
  minPlayers: number;
  indicator: OkeyTile | null;
  okey: OkeyTile | null;
  deck: OkeyTile[];
  discard: OkeyTile[];
  players: Okey101Player[];
  turnIndex: number;
  winnerId?: string;
  winnerName?: string;
  lastMove?: string;
  hostUserId: string;
}

export type Okey101Move =
  | { type: "start" }
  | { type: "draw" }
  | { type: "takeDiscard" }
  | { type: "discard"; tileId: string }
  | { type: "open" };

const COLORS: OkeyColor[] = ["red", "yellow", "blue", "black"];

export function createDeck(): OkeyTile[] {
  const tiles: OkeyTile[] = [];
  for (const color of COLORS) {
    for (let value = 1; value <= 13; value += 1) {
      tiles.push({ id: `${color}-${value}-a`, color, value });
      tiles.push({ id: `${color}-${value}-b`, color, value });
    }
  }
  tiles.push({ id: "joker-fake-a", color: "joker", value: 0, isFakeJoker: true });
  tiles.push({ id: "joker-fake-b", color: "joker", value: 0, isFakeJoker: true });
  return shuffle(tiles);
}

function shuffle<T>(arr: T[]): T[] {
  const copy = [...arr];
  for (let i = copy.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

function resolveOkey(indicator: OkeyTile): OkeyTile {
  if (indicator.isFakeJoker || indicator.color === "joker") {
    return { id: "okey-wild", color: "red", value: 1 };
  }
  const next = indicator.value === 13 ? 1 : indicator.value + 1;
  return { id: `okey-${indicator.color}-${next}`, color: indicator.color, value: next };
}

function isWild(tile: OkeyTile, okey: OkeyTile): boolean {
  if (tile.isFakeJoker || tile.color === "joker") return true;
  return tile.color === okey.color && tile.value === okey.value;
}

function tilePoints(tile: OkeyTile, okey: OkeyTile | null): number {
  if (isWild(tile, okey ?? { id: "", color: "red", value: 0 })) return 10;
  return tile.value;
}

function handPoints(hand: OkeyTile[], okey: OkeyTile | null): number {
  return hand.reduce((sum, t) => sum + tilePoints(t, okey), 0);
}

function hasBasicGroup(hand: OkeyTile[], okey: OkeyTile | null): boolean {
  const byColor = new Map<string, number[]>();
  const byValue = new Map<number, number>();
  for (const tile of hand) {
    const color = isWild(tile, okey!) ? "wild" : tile.color;
    const value = isWild(tile, okey!) ? -1 : tile.value;
    if (color !== "wild") {
      const list = byColor.get(color) ?? [];
      list.push(value);
      byColor.set(color, list);
    }
    if (value > 0) {
      byValue.set(value, (byValue.get(value) ?? 0) + 1);
    }
  }
  for (const values of byColor.values()) {
    values.sort((a, b) => a - b);
    let run = 1;
    for (let i = 1; i < values.length; i += 1) {
      if (values[i] === values[i - 1] + 1) {
        run += 1;
        if (run >= 3) return true;
      } else if (values[i] !== values[i - 1]) {
        run = 1;
      }
    }
  }
  for (const count of byValue.values()) {
    if (count >= 3) return true;
  }
  return hand.filter((t) => isWild(t, okey!)).length >= 2;
}

export function createOkey101State(hostUserId: string, hostName: string): Okey101State {
  return {
    gameType: "okey101",
    phase: "waiting",
    status: "waiting",
    openScoreTarget: 101,
    maxPlayers: 4,
    minPlayers: 2,
    indicator: null,
    okey: null,
    deck: [],
    discard: [],
    players: [
      {
        userId: hostUserId,
        displayName: hostName,
        seat: 0,
        hand: [],
        opened: false,
        penalty: 0,
        connected: true,
      },
    ],
    turnIndex: 0,
    hostUserId,
  };
}

export function addPlayer(
  state: Okey101State,
  userId: string,
  displayName: string,
): Okey101State {
  if (state.phase !== "waiting") throw new Error("Oyun başlamış");
  if (state.players.some((p) => p.userId === userId)) return state;
  if (state.players.length >= state.maxPlayers) throw new Error("Oda dolu");
  return {
    ...state,
    players: [
      ...state.players,
      {
        userId,
        displayName,
        seat: state.players.length,
        hand: [],
        opened: false,
        penalty: 0,
        connected: true,
      },
    ],
  };
}

function currentPlayer(state: Okey101State): Okey101Player {
  return state.players[state.turnIndex];
}

function nextTurnIndex(state: Okey101State): number {
  return (state.turnIndex + 1) % state.players.length;
}

export function startGame(state: Okey101State, userId: string): Okey101State {
  if (state.phase !== "waiting") throw new Error("Oyun zaten başladı");
  if (state.players.length < state.minPlayers) {
    throw new Error(`En az ${state.minPlayers} oyuncu gerekli`);
  }
  if (userId !== state.hostUserId) throw new Error("Yalnızca oda sahibi başlatabilir");

  let deck = createDeck();
  const indicator = deck.pop()!;
  const okey = resolveOkey(indicator);

  const players = state.players.map((p, index) => {
    const count = index === 0 ? 15 : 14;
    const hand: OkeyTile[] = [];
    for (let i = 0; i < count; i += 1) {
      const tile = deck.pop();
      if (tile) hand.push(tile);
    }
    return { ...p, hand, opened: false };
  });

  const firstDiscard = deck.pop();
  const discard = firstDiscard ? [firstDiscard] : [];

  return {
    ...state,
    phase: "playing",
    status: "playing",
    indicator,
    okey,
    deck,
    discard,
    players,
    turnIndex: 0,
    lastMove: "Oyun başladı",
  };
}

export function applyMove(
  state: Okey101State,
  userId: string,
  move: Okey101Move,
): Okey101State {
  if (move.type === "start") return startGame(state, userId);
  if (state.phase !== "playing") throw new Error("Oyun aktif değil");

  const playerIndex = state.players.findIndex((p) => p.userId === userId);
  if (playerIndex < 0) throw new Error("Oyuncu değilsiniz");
  if (playerIndex !== state.turnIndex) throw new Error("Sıra sizde değil");

  const player = state.players[playerIndex];
  const okey = state.okey!;

  switch (move.type) {
    case "draw": {
      if (player.drewThisTurn) throw new Error("Bu turda zaten çektiniz");
      if (state.deck.length === 0) throw new Error("Deste bitti");
      const tile = state.deck.pop()!;
      const players = [...state.players];
      players[playerIndex] = {
        ...player,
        hand: [...player.hand, tile],
        drewThisTurn: true,
      };
      return {
        ...state,
        deck: [...state.deck],
        players,
        lastMove: `${player.displayName} desteden çekti`,
      };
    }
    case "takeDiscard": {
      if (player.drewThisTurn) throw new Error("Bu turda zaten çektiniz");
      if (state.discard.length === 0) throw new Error("Atık yok");
      const tile = state.discard.pop()!;
      const players = [...state.players];
      players[playerIndex] = {
        ...player,
        hand: [...player.hand, tile],
        drewThisTurn: true,
      };
      return {
        ...state,
        discard: [...state.discard],
        players,
        lastMove: `${player.displayName} atılan taşı aldı`,
      };
    }
    case "discard": {
      if (!player.drewThisTurn && player.hand.length > 14) {
        throw new Error("Önce taş çekmelisiniz");
      }
      const idx = player.hand.findIndex((t) => t.id === move.tileId);
      if (idx < 0) throw new Error("Taş elinizde yok");
      const tile = player.hand[idx];
      const newHand = player.hand.filter((t) => t.id !== move.tileId);
      const players = [...state.players];
      players[playerIndex] = { ...player, hand: newHand };

      let next: Okey101State = {
        ...state,
        discard: [...state.discard, tile],
        players,
        lastMove: `${player.displayName} taş attı`,
      };

      if (player.opened && newHand.length === 0) {
        next = {
          ...next,
          phase: "finished",
          status: "finished",
          winnerId: userId,
          winnerName: player.displayName,
          turnIndex: playerIndex,
          lastMove: `${player.displayName} okey attı!`,
        };
        return applyPenalties(next, userId);
      }

      const nextIndex = nextTurnIndex(next);
      return {
        ...next,
        turnIndex: nextIndex,
        players: next.players.map((p) => ({ ...p, drewThisTurn: false })),
      };
    }
    case "open": {
      const points = handPoints(player.hand, okey);
      if (points < state.openScoreTarget) {
        throw new Error(`Açmak için en az ${state.openScoreTarget} puan gerekli`);
      }
      if (!hasBasicGroup(player.hand, okey)) {
        throw new Error("Geçerli per/çift yok");
      }
      const players = [...state.players];
      players[playerIndex] = { ...player, opened: true };
      return {
        ...state,
        players,
        lastMove: `${player.displayName} 101 ile açtı (${points} puan)`,
      };
    }
    default:
      throw new Error("Geçersiz hamle");
  }
}

function applyPenalties(state: Okey101State, winnerId: string): Okey101State {
  const okey = state.okey;
  const players = state.players.map((p) => {
    if (p.userId === winnerId) return p;
    const penalty = handPoints(p.hand, okey);
    return { ...p, penalty: p.penalty + penalty };
  });
  return { ...state, players };
}

export function publicView(state: Okey101State, viewerUserId: string) {
  return {
    gameType: state.gameType,
    phase: state.phase,
    status: state.status,
    openScoreTarget: state.openScoreTarget,
    maxPlayers: state.maxPlayers,
    minPlayers: state.minPlayers,
    indicator: state.indicator,
    okey: state.okey,
    deckCount: state.deck.length,
    discardTop: state.discard[state.discard.length - 1] ?? null,
    discardCount: state.discard.length,
    turnUserId: state.players[state.turnIndex]?.userId ?? null,
    turnName: state.players[state.turnIndex]?.displayName ?? null,
    winnerId: state.winnerId,
    winnerName: state.winnerName,
    lastMove: state.lastMove,
    hostUserId: state.hostUserId,
    players: state.players.map((p) => ({
      userId: p.userId,
      displayName: p.displayName,
      seat: p.seat,
      handCount: p.hand.length,
      opened: p.opened,
      penalty: p.penalty,
      connected: p.connected,
      hand: p.userId === viewerUserId ? p.hand : [],
      isMe: p.userId === viewerUserId,
    })),
    myHandPoints:
      state.players.find((p) => p.userId === viewerUserId)?.hand != null
        ? handPoints(
            state.players.find((p) => p.userId === viewerUserId)!.hand,
            state.okey,
          )
        : 0,
  };
}
